File: frontend/src/api/supplier-catalog.ts
import api from "@/api/client";

export interface SupplierCatalogEntry {
  id: string;
  company_id: string;
  supplier_id: string;
  product_id: string;
  supplier_ref?: string;
  purchase_price: number;
  lead_time_days: number;
  min_order_qty: number;
  notes?: string;
  is_active: boolean;
  created_at?: string;
  updated_at?: string;
  // Denormalized
  product_code?: string;
  product_name?: string;
  product_sale_price?: number;
  product_category?: string;
}

export interface CatalogSearchResult {
  id: string;
  supplier_id: string;
  supplier_name: string;
  supplier_code: string;
  product_id: string;
  product_code: string;
  product_name: string;
  supplier_ref?: string;
  purchase_price: number;
  sale_price: number;
  margin_pct?: number;
  lead_time_days: number;
}

export interface SupplierCatalogEntryCreate {
  product_id: string;
  supplier_ref?: string;
  purchase_price?: number;
  lead_time_days?: number;
  min_order_qty?: number;
  notes?: string;
}

export interface SupplierCatalogEntryUpdate {
  supplier_ref?: string;
  purchase_price?: number;
  lead_time_days?: number;
  min_order_qty?: number;
  notes?: string;
  is_active?: boolean;
}

export const supplierCatalogApi = {
  list: (supplierId: string) =>
    api.get<SupplierCatalogEntry[]>(`/suppliers/${supplierId}/catalog`).then((r) => r.data),

  add: (supplierId: string, data: SupplierCatalogEntryCreate) =>
    api.post<SupplierCatalogEntry>(`/suppliers/${supplierId}/catalog`, data).then((r) => r.data),

  update: (supplierId: string, entryId: string, data: SupplierCatalogEntryUpdate) =>
    api.put<SupplierCatalogEntry>(`/suppliers/${supplierId}/catalog/${entryId}`, data).then((r) => r.data),

  remove: (supplierId: string, entryId: string) =>
    api.delete(`/suppliers/${supplierId}/catalog/${entryId}`),

  search: (params?: { q?: string; supplier_id?: string; limit?: number }) =>
    api.get<CatalogSearchResult[]>("/catalog/search", { params }).then((r) => r.data),
};

File: backend/app/schemas/supplier_catalog.py
"""
SupplierProduct (catalog) schemas.
"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from uuid import UUID
from app.schemas.base import BaseSchema


class SupplierProductBase(BaseModel):
    """Base catalog entry fields."""
    product_id: UUID
    supplier_ref: Optional[str] = Field(None, max_length=50)
    purchase_price: float = Field(0.0, ge=0)
    lead_time_days: int = Field(0, ge=0)
    min_order_qty: float = Field(1.0, gt=0)
    notes: Optional[str] = None
    is_active: bool = True


class SupplierProductCreate(SupplierProductBase):
    """Create a catalog entry for a supplier."""
    pass


class SupplierProductUpdate(BaseModel):
    """Update a catalog entry — all fields optional."""
    supplier_ref: Optional[str] = Field(None, max_length=50)
    purchase_price: Optional[float] = Field(None, ge=0)
    lead_time_days: Optional[int] = Field(None, ge=0)
    min_order_qty: Optional[float] = Field(None, gt=0)
    notes: Optional[str] = None
    is_active: Optional[bool] = None


class SupplierProductResponse(BaseSchema):
    """Catalog entry response with product details."""
    id: UUID
    company_id: UUID
    supplier_id: UUID
    product_id: UUID
    supplier_ref: Optional[str] = None
    purchase_price: float
    lead_time_days: int
    min_order_qty: float
    notes: Optional[str] = None
    is_active: bool
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Denormalized product fields for display
    product_code: Optional[str] = None
    product_name: Optional[str] = None
    product_sale_price: Optional[float] = None
    product_category: Optional[str] = None


class CatalogSearchResult(BaseSchema):
    """Result from global catalog search across all suppliers."""
    id: UUID
    supplier_id: UUID
    supplier_name: str
    supplier_code: str
    product_id: UUID
    product_code: str
    product_name: str
    supplier_ref: Optional[str] = None
    purchase_price: float
    sale_price: float
    margin_pct: Optional[float] = None  # (sale - purchase) / purchase * 100
    lead_time_days: int
